design-proposals/guest-agent-exec.md

```diff
@@ -0,0 +1,56 @@
+## Virtctl Guest Agent Exec (ga-exec)
+
+**Authors:** Ryan Hallisey NVIDIA
+
+## Motivation
+The guest-agent is a powerful tool for querying a guest, but it isn’t easy for
+a user to do from kubectl.  To run a guest-agent command, the user has to locate
+a launcher pod, exec into the pod, run the guest-agent-exec command, capture the pid
+from the result, and run a second command - guest-agent-status - to get the result.  This
+process can be simplified.
+
+KubeVirt already has a lot of code using the guest-agent, so the idea for this proposal is
+to extend the existing connections to the guest-agent all the way up to the virtctl client
+so users can have ease of access too.
+
+#### Goals
+* Provide an client command for the user to run qemu guest-agent commands and return the output to stdout
+* Reduce the number of client commands it takes to run a guest-agent command
+
+#### Non-Goals
+* Allowing a user to open a shell in a guest
+
+#### Use-cases
+* As a user, I want to be able to run commands inside a guest and view the output in stout
+* As a user, I want to be able to view cloud-init logs with 1 command
+
+### Design
+KubeVirt already has the necessary functions in place to run the guest-agent exec
+command in the virt-hander and the virt-launcher, but those functions need to be
+exposed through the virt-api and virtctl in order to implement this feature.
+
+1. Create a ga-exec subresource
+2. Create a rest call in virt-handler
+3. Wire up the rest call to `Exec` client call (equivalent to `guest-exec` https://qemu.readthedocs.io/en/latest/interop/qemu-ga-ref.html#qapidoc-205)
+4. Create an abstraction that runs an `Exec` call to return the output from the first call (equivalent to `guest-exec-status` with the pid from 3. https://qemu.readthedocs.io/en/latest/interop/qemu-ga-ref.html#qapidoc-198)
+5. Create `virtctl ga-exec <command>`
+
+Example:
+Before adding `virtctl ga-exec`:
+```bash
+$ kubectl exec -it pod virt-launcher-123 - /bin/bash
+$ virsh qemu-agent-command 1 '{"execute": "guest-exec", "arguments":{"capture-output": true, "path":"powershell.exe", "arg": ["type", "\\user\\logs\\cloudinit\\cloudinit.log"]}}'`
+1234
+$ virsh qemu-agent-command 1 '{"execute": "guest-exec-status", "arguments":{"pid": 1234}}'`
+{result:somebase64string}
+```
+
+With `virtctl ga-exec`:
```

Comment:
I would try to iterate on the name a little bit. A few ideas:

 ```
 virtctl qemu-agent-exec ...
 virtchl qemu-exec ...
 virtctl guest-agent exec ...
 vrtctl qemu-guest-agent exec ...
 ```

Reply:
>virtctl guest-agent exec ...
>vrtctl qemu-guest-agent exec ...
>
What commands other than exec did you have in mind?

Reply:
I think exec makes a lot of sense, given that it closely resembles `kubect exec`. I just think that we don't have to work with abbreviations like `ga-` in a command. We have the opportunity of subcommand grouping.

Reply:
There would potentially already be at least three other commands which would make sense to be grouped under a `guest-agent` subcommand:

```
		vm.NewGuestOsInfoCommand(clientConfig),
		vm.NewUserListCommand(clientConfig),
		vm.NewFSListCommand(clientConfig),
```

Reply:
Sure, I understand not wanting to abbreviate.  So maybe `virtctl guest-agent-exec`?

If we did use the subcommand grouping, with `virtctl guest-agent exec`, I guess it allows us to expand the subgroup.  So I suppose it could support other guest agents?

Reply:
The `virtctl guest-agent exec` sub grouping is kind of interesting. I do like that it gives us more possibilities for additional ga commands in the future.

One thing I'm uncertain about is whether our grouping should involve the term `qemu` so it's clear which agent is being invoked. It seems possible that we might have additional guest agents in the future.

Reply:
Right, there will need to be a way to distinguish.  I guess `$ virtctl guest-agent <guest_agent> <command>`, so that would look like: `$ virtctl guest-agent qemu exec`.